MSVC 16.2, C++17, and /permissive- fails to build

Here is a minimal repro: https://github.com/wesleygriffin/absl-msvc-16-2

I don't see any duplicate issues opened or closed., although perhaps #356 looks related?

I compile with `/permissive-` to put MSVC into standards compliance mode: https://docs.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=vs-2019 as my main project is cross-platform.

The failure I'm getting is in absl/container/internal/compressed_tuple.h:

~~~
C:\Users\WesleyGriffin\Active\wesleygriffin\absl-msvc-16-2\out\build\x64-Debug\_deps\abseil-src\absl\container\internal\compressed_tuple.h(224): error C3546: '...': there are no parameter packs available to expand
  C:\Users\WesleyGriffin\Active\wesleygriffin\absl-msvc-16-2\out\build\x64-Debug\_deps\abseil-src\absl\container\internal\inlined_vector.h(436): note: see reference to class template instantiation 'absl::container_internal::CompressedTuple<std::allocator<T>,unsigned __int64>' being compiled
          with
          [
              T=Foo
          ]
  C:\Users\WesleyGriffin\Active\wesleygriffin\absl-msvc-16-2\out\build\x64-Debug\_deps\abseil-src\absl\container\inlined_vector.h(73): note: see reference to class template instantiation 'absl::inlined_vector_internal::Storage<T,8,A>' being compiled
          with
          [
              T=Foo,
              A=std::allocator<Foo>
          ]
  C:\Users\WesleyGriffin\Active\wesleygriffin\absl-msvc-16-2\test.cpp(10): note: see reference to class template instantiation 'absl::InlinedVector<Foo,8,std::allocator<T>>' being compiled
          with
          [
              T=Foo
          ]
C:\Users\WesleyGriffin\Active\wesleygriffin\absl-msvc-16-2\out\build\x64-Debug\_deps\abseil-src\absl\container\internal\compressed_tuple.h(228): error C3520: 'Vs': parameter pack must be expanded in this context
~~~

If I don't use `/permissive-` then the repro compiles cleanly.

Same here, basic use of `absl::flat_hash_map` in [VS 2019](https://visualstudio.microsoft.com/vs/) results in compilation errors:

> absl\container\internal\compressed_tuple.h(225,54): error C3546: '...': there are no parameter packs available to expand

The problem was introduced by c6c3c1b49.

Removing `/permissive-` makes VS2019 compile the code, but it doesn't link because apparently abseil containers aren't header-only (unlike boost::container::flat_map, flat_hash::hash_map and other libs from the web).
**Edit:** Just tried to make it compile with abseil, and I had to add 32 .cc files to my project. I'm shocked

--- a/absl/container/internal/compressed_tuple.h
+++ b/absl/container/internal/compressed_tuple.h
@@ -169,9 +169,28 @@ constexpr bool ShouldAnyUseBase() {
 }
 
 template <typename T, typename V>
-using TupleMoveConstructible = typename std::conditional<
+using TupleElementMoveConstructible = typename std::conditional<
       std::is_reference<T>::value, std::is_convertible<V, T>,
       std::is_constructible<T, V&&>>::type;
+
+template <bool SizeMatches, class T, class... Vs>
+struct TupleMoveConstructible : std::false_type {};
+
+template <class... Ts, class... Vs>
+struct TupleMoveConstructible<true, CompressedTuple<Ts...>, Vs...>
+    : std::integral_constant<bool, absl::conjunction<TupleElementMoveConstructible<Ts, Vs&&>...>::value> {
+};
+
+template<typename... Es>
+struct compressed_tuple_size;
+
+template<typename... Es>
+struct compressed_tuple_size<CompressedTuple<Es...>>
+    : public std::integral_constant<std::size_t, sizeof...(Es)> {};
+
+template <class T, class... Vs>
+struct TupleItemsMoveConstructible
+    : std::integral_constant<bool, TupleMoveConstructible<compressed_tuple_size<T>::value == sizeof...(Vs), T, Vs...>::value> {};
 
 }  // namespace internal_compressed_tuple
 
@@ -217,17 +236,18 @@ class ABSL_INTERNAL_COMPRESSED_TUPLE_DECLSPEC CompressedTuple
   explicit constexpr CompressedTuple(const Ts&... base)
       : CompressedTuple::CompressedTupleImpl(absl::in_place, base...) {}
 
-  template <typename... Vs,
+  template <typename First, typename... Vs,
             absl::enable_if_t<
                 absl::conjunction<
                     // Ensure we are not hiding default copy/move constructors.
                     absl::negation<std::is_same<void(CompressedTuple),
-                                                void(absl::decay_t<Vs>...)>>,
-                    internal_compressed_tuple::TupleMoveConstructible<
-                        Ts, Vs&&>...>::value,
+                                                void(absl::decay_t<First>)>>,
+                    internal_compressed_tuple::TupleItemsMoveConstructible<
+                        CompressedTuple<Ts...>, First, Vs...>>::value,
                 bool> = true>
-  explicit constexpr CompressedTuple(Vs&&... base)
+  explicit constexpr CompressedTuple(First&& first, Vs&&... base)
       : CompressedTuple::CompressedTupleImpl(absl::in_place,
+                                             absl::forward<First>(first),
                                              absl::forward<Vs>(base)...) {}
 
   template <int I>

--- a/absl/container/internal/compressed_tuple.h
+++ b/absl/container/internal/compressed_tuple.h
@@ -169,28 +169,33 @@ constexpr bool ShouldAnyUseBase() {
 }
 
 template <typename T, typename V>
-using TupleElementMoveConstructible = typename std::conditional<
-      std::is_reference<T>::value, std::is_convertible<V, T>,
-      std::is_constructible<T, V&&>>::type;
+using TupleElementMoveConstructible =
+    typename std::conditional<std::is_reference<T>::value,
+                              std::is_convertible<V, T>,
+                              std::is_constructible<T, V&&>>::type;
 
 template <bool SizeMatches, class T, class... Vs>
 struct TupleMoveConstructible : std::false_type {};
 
 template <class... Ts, class... Vs>
 struct TupleMoveConstructible<true, CompressedTuple<Ts...>, Vs...>
-    : std::integral_constant<bool, absl::conjunction<TupleElementMoveConstructible<Ts, Vs&&>...>::value> {
-};
+    : std::integral_constant<
+          bool, absl::conjunction<
+                    TupleElementMoveConstructible<Ts, Vs&&>...>::value> {};
 
-template<typename... Es>
+template <typename T>
 struct compressed_tuple_size;
 
-template<typename... Es>
+template <typename... Es>
 struct compressed_tuple_size<CompressedTuple<Es...>>
     : public std::integral_constant<std::size_t, sizeof...(Es)> {};
 
 template <class T, class... Vs>
 struct TupleItemsMoveConstructible
-    : std::integral_constant<bool, TupleMoveConstructible<compressed_tuple_size<T>::value == sizeof...(Vs), T, Vs...>::value> {};
+    : std::integral_constant<
+          bool, TupleMoveConstructible<compressed_tuple_size<T>::value ==
+                                           sizeof...(Vs),
+                                       T, Vs...>::value> {};
 
 }  // namespace internal_compressed_tuple
 

